src/compositemark/errorbar.ts

```diff
@@ -1,42 +1,395 @@
+      } else if (encoding.x.aggregate === ERRORBAR && encoding.y.aggregate === ERRORBAR) {
+        throw new Error('Both x and y cannot have aggregate');
+      } else {
+        if (isMarkDef(mark) && mark.orient) {
+          return mark.orient;
+        }
+
+        // default orientation = vertical
+        return 'vertical';
+      }
+    }
+
+    // x is continuous but y is not
+    return 'horizontal';
+  } else if (isFieldDef(encoding.y) && isContinuous(encoding.y)) {
+    // y is continuous but x is not
+    return 'vertical';
+  } else {
+    // Neither x nor y is continuous.
+    throw new Error('Need a valid continuous axis for errorbars');
+  }
+}
+
+
+function errorBarContinousAxis(spec: GenericUnitSpec<Encoding<string>, ErrorBar | ErrorBarDef>, orient: Orient) {
+  const {mark: mark, encoding: encoding, projection: _p, ..._outerSpec} = spec;
+
+  let continuousAxisChannelDef: PositionFieldDef<string>;
+  let continuousAxis: 'x' | 'y';
+
+  if (orient === 'vertical') {
+    continuousAxis = 'y';
+    continuousAxisChannelDef = encoding.y as FieldDef<string>; // Safe to cast because if y is not continuous fielddef, the orient would not be vertical.
+  } else {
+    continuousAxis = 'x';
+    continuousAxisChannelDef = encoding.x as FieldDef<string>; // Safe to cast because if x is not continuous fielddef, the orient would not be horizontal.
+  }
+
+  if (continuousAxisChannelDef && continuousAxisChannelDef.aggregate) {
+    const {aggregate, ...continuousAxisWithoutAggregate} = continuousAxisChannelDef;
+    if (aggregate !== ERRORBAR) {
+      log.warn(`Continuous axis should not have customized aggregation function ${aggregate}`);
+    }
+    continuousAxisChannelDef = continuousAxisWithoutAggregate;
+  }
+
+  return {
+    continuousAxisChannelDef,
+    continuousAxis
+  };
+}
+
+function isExtentCI(extent: ErrorBarExtent): extent is 'ci' {
+  return extent === 'ci';
+}
+
+function isExtentIqr(extent: ErrorBarExtent): extent is 'iqr' {
+  return extent === 'iqr';
+}
+
+function errorBarSpecialExtents(extent: 'ci' | 'iqr', fieldName: string): AggregatedFieldDef[] {
+  return [
+    {
+      op: (extent === 'ci') ? 'ci0' : 'q1',
+      field: fieldName,
+      as: 'lower_rule_' + fieldName
+    },
+    {
+      op: (extent === 'ci') ? 'ci1' : 'q3',
+      field: fieldName,
+      as: 'upper_rule_' + fieldName
+    }
+  ];
+}
+
+function errorBarParams(spec: GenericUnitSpec<Encoding<string>, ErrorBar | ErrorBarDef>, orient: Orient, center: ErrorBarCenter, extent: ErrorBarExtent) {
+
+  const {continuousAxisChannelDef, continuousAxis} = errorBarContinousAxis(spec, orient);
+  const encoding = spec.encoding;
+
+  const continuousFieldName: string = continuousAxisChannelDef.field;
+
+  let aggregate: AggregatedFieldDef[] = [
+    {
+      op: center,
+      field: continuousFieldName,
+      as: center + '_' + continuousFieldName
+    }
+  ];
+
+  let postAggregateCalculates: CalculateTransform[] = [];
+
+  if (isExtentCI(extent)) {
```

Comment:
You don't need two separate conditions for CI and IQR.  

`if (extent === 'ci' || extent === 'iqr')`  is sufficient. 

You can also just remove the `errorBarSpecialExtents` method and move its logic here.